internal/server/data/data.go

```diff
@@ -152,28 +139,62 @@ func add[T models.Modelable](db *gorm.DB, model *T) error {
 		return err
 	}
 
-	if err := db.Create(model).Error; err != nil {
-		if isUniqueConstraintViolation(err) {
-			return fmt.Errorf("%w: %s", internal.ErrDuplicate, err)
-		}
+	err := db.Create(model).Error
+	return handleError(err)
+}
 
-		return err
-	}
+type UniqueConstraintError struct {
+	Table  string
+	Column string
+}
 
-	return nil
+func (e UniqueConstraintError) Error() string {
+	return fmt.Sprintf("value for %v already exists in %v", e.Column, e.Table)
 }
 
-func isUniqueConstraintViolation(err error) bool {
-	var pgerr *pgconn.PgError
-	if errors.As(err, &pgerr) {
-		return pgerr.Code == pgerrcode.UniqueViolation
+// handleError looks for well known DB errors. If the error is recognized it
+// is translated into a UniqueConstraintError so that calling code can
+// inspect the error.
+func handleError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var pgErr *pgconn.PgError
+	if errors.As(err, &pgErr) {
+		switch pgErr.Code {
+		case pgerrcode.UniqueViolation:
+			// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/access/nbtree/nbtinsert.c;h=f6f4af8bfe3f0876e860944e2233328f8cc3303a;hb=HEAD#l668
+			// ereport(ERROR,
+			//         (errcode(ERRCODE_UNIQUE_VIOLATION),
+			//          errmsg("duplicate key value violates unique constraint \"%s\"",
+			//                 RelationGetRelationName(rel)),
+			//                 key_desc ? errdetail("Key %s already exists.",
+			//                            key_desc) : 0,
+			//                 errtableconstraint(heapRel,
+			//                                    RelationGetRelationName(rel))));
+			re := regexp.MustCompile(`Key \((?P<columnName>[[:print:]]+)\)=\([[:print:]]+\) already exists.`)
+			matches := re.FindStringSubmatch(pgErr.Detail)
+			if matches != nil {
+				columnNameIndex := re.SubexpIndex("columnName")
+				return UniqueConstraintError{Table: pgErr.TableName, Column: matches[columnNameIndex]}
+			}
+		}
 	}
 
-	if strings.HasPrefix(err.Error(), "UNIQUE constraint failed:") {
-		return true
+	// https://sqlite.org/src/file?name=ext/rtree/rtree.c:
+	// pRtree->base.zErrMsg = sqlite3_mprintf(
+	//     "UNIQUE constraint failed: %s.%s", pRtree->zName, zCol
+	// );
+	re := regexp.MustCompile(`UNIQUE constraint failed: (?P<tableName>[[:print:]]+)\.(?P<columnName>[[:print:]]+)`)
+	matches := re.FindStringSubmatch(err.Error())
+	if matches != nil {
+		tableNameIndex := re.SubexpIndex("tableName")
```

Comment:
I feel the same way about regex so I'm happy to look for an alternative. In fact, the first iterations used split/cut which works fine for sqlite but gets messy for postgres. Ideally, both variants use the same type of code for consistency and simplicity. 

`handleError` can take the table as an input since the caller has the model but this isn't really the problem. The table name is readily available for postgres and parsable for sqlite, which we need to do anyways. Additionally, I don't believe we're using any implicit associations so the model passed in should mirror the table name set by the error.

Reply:
I'm trying `strings.FieldsFunc` which is nicer is some areas but more brittle in others. Let me know what you think